### PR TITLE
Add action post, post-if support

### LINKED ISSUE #626: Issue: "post" run not executed
## Act version

```none
act version 0.2.21
```

## Expected behaviour

After executing the steps defined in the workflow, it should run the "post" runs, here specifically demonstrated for `actions/checkout` it should say things like "Post job cleanup.", as can be seen [here](https://github.com/sverhagen/act-post-run/runs/2388459404) and in the screenshot:

image

For the example here (using `actions/checkout`), the "post" run is defined [here](https://github.com/actions/checkout/blob/25a956c84d5dd820d28caab9f86b8d183aeeff3d/action.yml#L74), and implemented [here](https://github.com/actions/checkout/blob/25a956c84d5dd820d28caab9f86b8d183aeeff3d/dist/index.js#L3302).

## Actual behaviour

The "post" run for `actions/checkout` is skipped. Seen the same for other actions (not in this example), such as `actions/cache`.

## Workflow and/or repository

<details>
  <summary>Workflow</summary>

```none
name: Build

on: [push]

jobs:
  build_job:
    name: Build
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - run: echo "Test"
```
</details>

## Steps to reproduce

1. Clone example repo (https://github.com/sverhagen/act-post-run)
2. Enter cloned repo directory
3. Run `act`:
    ```
    act
    ```

## `act` output

<details>
  <summary>Log</summary>

```none
DEBU[0000] Loading environment from /home/sander/Code/act-test/.env 
DEBU[0000] Loading secrets from /home/sander/Code/act-test/.secrets 
DEBU[0000] Loading workflows from '/home/sander/Code/act-test/.github/workflows' 
DEBU[0000] Reading workflow '/home/sander/Code/act-test/.github/workflows/build.yaml' 
DEBU[0000] Planning event: push                         
DEBU[0000] Loading slug from git directory '/home/sander/Code/act-test/.git' 
DEBU[0000] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0000] Loading revision from git directory '/home/sander/Code/act-test/.git' 
DEBU[0000] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0000] HEAD points to '6d16831619f89423bb42e580f31ea9be64114946' 
DEBU[0000] HEAD matches refs/heads/main                 
DEBU[0000] using github ref: refs/heads/main            
DEBU[0000] context env => map[ACT:true]                 
[Build/Build] 🚀  Start image=catthehacker/ubuntu:act-latest
DEBU[0000] Loading slug from git directory '/home/sander/Code/act-test/.git' 
DEBU[0000] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0000] Loading revision from git directory '/home/sander/Code/act-test/.git' 
DEBU[0000] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0000] HEAD points to '6d16831619f89423bb42e580f31ea9be64114946' 
DEBU[0000] HEAD matches refs/heads/main                 
DEBU[0000] using github ref: refs/heads/main            
[Build/Build]   🐳  docker pull catthehacker/ubuntu:act-latest
DEBU[0000] Image exists? true                           
[Build/Build]   🐳  docker create image=catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[Build/Build] Created container name=act-Build-Build id=9652f64e42fdf2f2b2922416ea491edb2c824ac061ea806e40c7ff38a580147d from image catthehacker/ubuntu:act-latest (platform: linux/amd64)
[Build/Build] ENV ==> [RUNNER_TOOL_CACHE=/opt/hostedtoolcache RUNNER_OS=Linux RUNNER_TEMP=/tmp]
[Build/Build]   🐳  docker run image=catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[Build/Build] Starting container: 9652f64e42fdf2f2b2922416ea491edb2c824ac061ea806e40c7ff38a580147d
[Build/Build] Started container: 9652f64e42fdf2f2b2922416ea491edb2c824ac061ea806e40c7ff38a580147d
[Build/Build]   🐳  docker cp src=/home/sander/Code/act-test/. dst=/home/sander/Code/act-test
[Build/Build] Exec command '[mkdir -p /home/sander/Code/act-test]'
DEBU[0000] Writing tarball /tmp/act916293007 from /home/sander/Code/act-test/. 
DEBU[0000] Stripping prefix:/home/sander/Code/act-test/ src:/home/sander/Code/act-test/. 
[Build/Build] Extracting content from '/tmp/act916293007' to '/home/sander/Code/act-test'
DEBU[0000] Writing entry to tarball workflow/event.json len:2 
DEBU[0000] Writing entry to tarball workflow/envs.txt len:0 
DEBU[0000] Writing entry to tarball home/.act len:0     
[Build/Build] Extracting content to '/home/sander/Code'
DEBU[0000] Loading slug from git directory '/home/sander/Code/act-test/.git' 
DEBU[0000] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0000] Loading revision from git directory '/home/sander/Code/act-test/.git' 
DEBU[0000] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0000] HEAD points to '6d16831619f89423bb42e580f31ea9be64114946' 
DEBU[0000] HEAD matches refs/heads/main                 
DEBU[0000] using github ref: refs/heads/main            
DEBU[0001] Loading slug from git directory '/home/sander/Code/act-test/.git' 
DEBU[0001] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0001] Loading revision from git directory '/home/sander/Code/act-test/.git' 
DEBU[0001] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0001] HEAD points to '6d16831619f89423bb42e580f31ea9be64114946' 
DEBU[0001] HEAD matches refs/heads/main                 
DEBU[0001] using github ref: refs/heads/main            
DEBU[0001] context env => map[ACT:true]                 
DEBU[0001] context env => map[ACT:true CI:true GITHUB_ACTION:0 GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/home/sander/Code/workflow/envs.txt GITHUB_EVENT_NAME:push GITHUB_EVENT_PATH:/home/sander/Code/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/main GITHUB_REPOSITORY:sverhagen/act-post-run GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:6d16831619f89423bb42e580f31ea9be64114946 GITHUB_TOKEN: GITHUB_WORKFLOW:Build GITHUB_WORKSPACE:/home/sander/Code/act-test HOME:/home/sander/Code/home ImageOS:ubuntu20] 
[Build/Build] setupEnv => map[ACT:true CI:true GITHUB_ACTION:0 GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/home/sander/Code/workflow/envs.txt GITHUB_EVENT_NAME:push GITHUB_EVENT_PATH:/home/sander/Code/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/main GITHUB_REPOSITORY:sverhagen/act-post-run GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:6d16831619f89423bb42e580f31ea9be64114946 GITHUB_TOKEN: GITHUB_WORKFLOW:Build GITHUB_WORKSPACE:/home/sander/Code/act-test HOME:/home/sander/Code/home ImageOS:ubuntu20]
[Build/Build] ⭐  Run actions/checkout@v2
DEBU[0001] Loading slug from git directory '/home/sander/Code/act-test/.git' 
DEBU[0001] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0001] Loading revision from git directory '/home/sander/Code/act-test/.git' 
DEBU[0001] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0001] HEAD points to '6d16831619f89423bb42e580f31ea9be64114946' 
DEBU[0001] HEAD matches refs/heads/main                 
DEBU[0001] using github ref: refs/heads/main            
[Build/Build] Skipping actions/checkout
[Build/Build]   ✅  Success - actions/checkout@v2
DEBU[0001] Loading slug from git directory '/home/sander/Code/act-test/.git' 
DEBU[0001] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0001] Loading revision from git directory '/home/sander/Code/act-test/.git' 
DEBU[0001] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0001] HEAD points to '6d16831619f89423bb42e580f31ea9be64114946' 
DEBU[0001] HEAD matches refs/heads/main                 
DEBU[0001] using github ref: refs/heads/main            
DEBU[0001] Loading slug from git directory '/home/sander/Code/act-test/.git' 
DEBU[0001] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0001] Loading revision from git directory '/home/sander/Code/act-test/.git' 
DEBU[0001] Found revision: 6d16831619f89423bb42e580f31ea9be64114946 
DEBU[0001] HEAD points to '6d16831619f89423bb42e580f31ea9be64114946' 
DEBU[0001] HEAD matches refs/heads/main                 
DEBU[0001] using github ref: refs/heads/main            
DEBU[0001] context env => map[ACT:true]                 
DEBU[0001] context env => map[ACT:true CI:true GITHUB_ACTION:1 GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/home/sander/Code/workflow/envs.txt GITHUB_EVENT_NAME:push GITHUB_EVENT_PATH:/home/sander/Code/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/main GITHUB_REPOSITORY:sverhagen/act-post-run GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:6d16831619f89423bb42e580f31ea9be64114946 GITHUB_TOKEN: GITHUB_WORKFLOW:Build GITHUB_WORKSPACE:/home/sander/Code/act-test HOME:/home/sander/Code/home ImageOS:ubuntu20] 
[Build/Build] setupEnv => map[ACT:true CI:true GITHUB_ACTION:1 GITHUB_ACTIONS:true GITHUB_ACTOR:nektos/act GITHUB_API_URL:https://api.github.com GITHUB_ENV:/home/sander/Code/workflow/envs.txt GITHUB_EVENT_NAME:push GITHUB_EVENT_PATH:/home/sander/Code/workflow/event.json GITHUB_GRAPHQL_URL:https://api.github.com/graphql GITHUB_REF:refs/heads/main GITHUB_REPOSITORY:sverhagen/act-post-run GITHUB_RUN_ID:1 GITHUB_RUN_NUMBER:1 GITHUB_SERVER_URL:https://github.com GITHUB_SHA:6d16831619f89423bb42e580f31ea9be64114946 GITHUB_TOKEN: GITHUB_WORKFLOW:Build GITHUB_WORKSPACE:/home/sander/Code/act-test HOME:/home/sander/Code/home ImageOS:ubuntu20]
[Build/Build] ⭐  Run echo "Test"
DEBU[0001] Wrote command '
echo "Test"
' to 'workflow/1' 
DEBU[0001] Writing entry to tarball workflow/1 len:11   
[Build/Build] Extracting content to '/home/sander/Code/'
[Build/Build] Exec command '[bash --login --noprofile --norc -e -o pipefail /home/sander/Code/workflow/1]'
| Test
[Build/Build]   ✅  Success - echo "Test"
[Build/Build] Removed container: 9652f64e42fdf2f2b2922416ea491edb2c824ac061ea806e40c7ff38a580147d
[Build/Build]   🐳  docker volume rm act-Build-Build
```
</details>

## Notes

"As seen in" [this Gitter question](https://gitter.im/nektos/act?at=607bf633b9e6de24d64173ba).

Great project, thank you.

### PR DESCRIPTION
**Pull request by sp-ricard-valverde**
fixes #626 

Support for running actions that define `post` steps with or without `post-if` conditions

Made changes to support dependency injection for modeling Actions by using and implementing ActionProviders, this allows for improved mocking and testing Actions

Because this PR was first aimed at giving support to https://github.com/actions/cache , a [simple local cache server is also provided in the form of a Docker image](https://github.com/sp-ricard-valverde/github-act-cache-server) (necessary for the cache action to work properly)

### CODE REVIEW

File: pkg/runner/provider.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+	"context"
+	"embed"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	// Go told me to?
```
Comment (catthehacker) [inline pkg/runner/provider.go:14]: `ctrl` + `f`
`path.Join`
😉

File: pkg/runner/runner_test.go [hunk 2]
```diff
@@ -269,3 +282,120 @@ func TestContainerPath(t *testing.T) {
...
+			}
+			runner, err := New(runnerConfig, providers)
+			assert.Nil(t, err, workflowPath)
+
+			planner, err := model.NewWorkflowPlanner(fullWorkflowPath, true)
+			assert.Nil(t, err, fullWorkflowPath)
+
+			plan := planner.PlanEvent(eventName)
+
+			err = runner.NewPlanExecutor(plan)(ctx)
```
Comment (catthehacker) [inline pkg/runner/runner_test.go:388]:
```suggestion
			err = runner.NewPlanExecutor(plan, providers)(ctx)
```

```diff
- func New(runnerConfig *Config) (Runner, error) {
+ func New(runnerConfig *Config, providers *Providers) (Runner, error) {
```

**Comment by sp-ricard-valverde** [comment]
Could the CI linter check error be wrong ?

> Restored cache for golangci-lint from key 'golangci-lint.cache-2706-f4b446ed140060ca871bb320af1bfb5a337ed1b2' in 2659ms
Prepared env in 2659ms
run golangci-lint
  Running [/home/runner/golangci-lint-1.40.0-linux-amd64/golangci-lint run --out-format=github-actions] in [] ...
  Error: too few arguments in call to runner.New (typecheck)
  Error: issues found

But it's already like this https://github.com/sp-ricard-valverde/act/blob/8a175b0910e331477bf3209a69eb6ec237262e42/cmd/root.go#L281 .
I've passed `golangci-lint 1.43.0`(the runner uses `1.40.0`) locally without problems. I believe the `golangci-lint` cache might be causing the issue here.

**Comment by sp-ricard-valverde** [comment]
And I believe the `test-linux` error check is about the 85% coverage, I'll try to cover those too.

**Comment by sp-ricard-valverde** [comment]
Ok I see I missed a recent rebase on `master`. Working on it !

**Comment by sp-ricard-valverde** [comment]
Done

**Comment by catthehacker** [comment]
It will take some time to review whole PR

**Comment by sp-ricard-valverde** [comment]
> It will take some time to review whole PR

Let me know if I can help clarifying or something else

**Review by catthehacker — APPROVED** [review]

**Comment by sp-ricard-valverde** [comment]
Closing as I don't have enough time to maintain such a big PR open for so long

**Comment by ZauberNerd** [comment]
@KnisterPeter and me will try to give this another go. Thanks for your work!

**Pull request closed without merging**